docs/architecture/adr-018-dynamic-capability-store.md

```diff
@@ -0,0 +1,197 @@
+  }
+}
+```
+
+The `CapabilityKeeper` also provides the ability to create *scoped* sub-keepers which are tied to a particular module name. These `ScopedCapabilityKeeper`s must be created at application
+initialisation and passed to modules, which can then use them to claim capabilities they receive and retrieve capabilities which they own by name.
+
+```golang
+type ScopedCapabilityKeeper struct {
+  capabilityKeeper CapabilityKeeper
+  moduleName string
+}
+```
+
+`ScopeToModule` is used to create a scoped sub-keeper with a particular name, which must be unique.
+
+```golang
+func (ck CapabilityKeeper) ScopeToModule (moduleName string) {
+  if _, present := ck.moduleNames[moduleName]; present {
+    panic("cannot create multiple scoped capability keepers for the same module name")
+  }
+  ck.moduleNames[moduleName] = interface{}
+  return ScopedCapabilityKeeper{
+    capabilityKeeper: ck,
+    moduleName: moduleName
+  }
+}
+```
+
+`ClaimCapability` allows a module to claim a capability key which it has received (perhaps by calling `NewCapability`, or from another module), so that future `GetCapability` calls will succeed.
+
+`ClaimCapability` MUST be called, even if `NewCapability` was called by the same module. Capabilities are single-owner, so if multiple modules have a single `Capability` reference, the last module
+to call `ClaimCapability` will own it. To avoid confusion, a module which calls `NewCapability` SHOULD either call `ClaimCapability` or pass the capability to another module which will then claim it.
+
+```golang
+func (sck ScopedCapabilityKeeper) ClaimCapability(ctx sdk.Context, capability Capability) error {
+  persistentStore := ctx.KVStore(sck.ck.persistentKey)
+  memoryStore := ctx.KVStore(sck.ck.memoryKey)
+  // fetch name from memory store
+  name := memoryStore.Get("fwd/" + capability)
+  if name === nil {
+    return errors.New("capability not found")
+  }
+  // set name to module in persistent store
+  persistentStore.Set(name, sck.moduleName)
+}
+```
+
+`GetCapability` allows a module to fetch a capability which it has previously claimed by name. The module is not allowed to retrieve capabilities which it does not own.
+
+```golang
+func (sck ScopedCapabilityKeeper) GetCapability(ctx sdk.Context, name string) (Capability, error) {
+  persistentStore := ctx.KVStore(sck.ck.persistentKey)
+  memoryStore := ctx.KVStore(sck.ck.memoryKey)
+  // check that this module owns the capability with this name
+  res := persistentStore.Get(name)
+  if res != sck.moduleName {
+    return errors.New("capability of this name not owned by module")
+  }
+  // fetch capability from memory store, return it
+  capability := memoryStore.Get("rev/" + name)
+  return capability
+}
+```
+
+### Memory store
+
+A new store type, `StoreTypeMemory`, will be added to the `store` package. A new store key type, `MemoryStoreKey`, will be added to the `store` package. The `MemoryStoreKey`s work just like `sdk.StoreKey`s.
```

Comment:
It might be obvious, but what advantage does a `StoreTypeMemory` have over each keeper simply keeping in-memory state in a `map` since the `StoreTypeMemory` is name-spaced anyway?

Reply:
The advantage is that it works with the context, and any sort of `sdk.CacheContext` usage, and doesn't break the "keepers don't store state" model.

Reply:
(that could have changed, if it has, we could use a map in the keeper instead)

Reply:
No, I think that's fair. Apart from cache, that model still holds true.